tensorpack/tfutils/scope_utils.py

```diff
@@ -31,18 +31,11 @@ def myfunc(x):
             myfunc(x3)  # will inherit parent scope reuse
             myfunc(x4)  # will reuse
     """
-    used_scope = set()
 
     @functools.wraps(func)
     def wrapper(*args, **kwargs):
         scope = tf.get_variable_scope()
-        h = hash((tf.get_default_graph(), scope.name))
-        # print("Entering " + scope.name + " reuse: " + str(h in used_scope))
-        if h in used_scope:
-            with tf.variable_scope(scope, reuse=True):
-                return func(*args, **kwargs)
-        else:
-            used_scope.add(h)
+        with tf.variable_scope(scope, reuse=tf.AUTO_REUSE):
```

Comment:
It's error-prone to put any user code under `tf.AUTO_REUSE`. Because there you don't know whether your scope is reuse or not -- it's both. So a user can end up creating unexpected variables. So I prefer to keep it as it is.
Also, AUTO_REUSE is a tf1.4 thing.

Reply:
But the current auto_reuse_version has some issues as well. It affects the entire var_scope, while `tf.AUTO_REUSE` seems to be per-variable.

Reply:
By design it was meant to affect the entire scope. There is a name confusion with `tf.AUTO_REUSE` because I write this function before AUTO_REUSE appears.

Reply:
Creating unexpected (new) variables under `reuse=True` leads to an error. But it will silently pass under `tf.AUTO_REUSE`. That's the kind of error I'm talking about

Reply:
The `tf.AUTO_REUSE` removes workarounds like:
https://github.com/cgtuebingen/learning-blind-motion-deblurring/blob/master/learning_blind_motion_deblurring.py#L41-L47

Reply:
I know `tf.AUTO_REUSE` is useful in many situations. But because it's also error-prone at the same time, I would let users to decide to use it.

Reply:
Many tf stuff (EMA, tf.layers) create variables secretly that I wouldn't have known them until I saw the error under a reuse scope.